sphinx/ext/autodoc/__init__.py

```diff
@@ -899,6 +902,15 @@ def generate(self, more_content: Any = None, real_modname: str = None,
             except PycodeError:
                 pass
 
+        try:
+            subject = inspect.unpartial(self.object)
+            self.canonical_objpath = subject.__qualname__.split(".")
+            self.canonical_fullname = '%s::%s' % (self.real_modname, subject.__qualname__)
```

Comment:
Could you add an initialization of this member to `__init__()` and fill it in `parse_name()`? It would be better to keep this method simple as possible.

Reply:
I used a property.  Using a property feels safer than hoping for users to have `parse_name` used first.  Actually, I'm wondering why there is so little use of properties for lazy loaded attributes